htsinfer/get_read_layout.py

```diff
@@ -0,0 +1,286 @@
+
+        records: int = 0
+        total_count: int = 0
+
+        try:
+            with open(self.path) as _f:  # type: ignore
+
+                LOGGER.debug("Procecssing Reads")
+                try:
+                    for record in FastqGeneralIterator(source=_f):
+                        read = record[1]
+
+                        # Searching for adapters in read
+                        for _, (_, key) in self.trie.iter(read):
+                            self.adapter_counts[key] += 1
+                            total_count += 1
+
+                        records += 1
+
+                except Exception as exc:
+                    self.result = None
+                    raise FileProblem(f"File is corrupt: {self.path}") from exc
+        except (OSError, ValueError) as exc:
+            self.result = None
+            raise FileProblem(f"{type(exc).__name__}: {str(exc)}") from exc
+
+        LOGGER.debug(f"Total records processed: {records}")
+
+        if total_count != 0:
+            # process and convert data to data frame
+            for i in self.adapter_counts:
+                self.adapter_counts[i] = round(
+                    (self.adapter_counts[i]/records)*100, 2
+                    )
+            adapters_df = self._convert_dict_to_df()
+
+            # write data frame (in JSON) to file
+            name = (
+                Path(self.out_dir) / f"read_layout_{Path(self.path).name}.json"
+            )
+            LOGGER.debug(f"Writing results to file: {name}")
+            adapters_df.to_json(
+                name,
+                orient='split',
+                index=False,
+                indent=True,
+            )
+
+            # validate results
+            if results_validator(
+                data=adapters_df,
+                column_index=1,
+                min_value=self.min_match_pct,
+                min_ratio=self.min_freq_ratio,
+            ):
+                self.result = adapters_df.iloc[0]['Adapter Sequence']
+            else:
+                self.result = None
+        else:
+            self.result = None
+
+    def _convert_dict_to_df(self) -> DataFrame:
+        """Convert dictionary to data frame.
+
+        Returns:
+            Data frame of adapter sequences and corresponding count
+            percentages.
+        """
+        adapters_df = DataFrame(self.adapter_counts.items())
+        adapters_df.columns = ['Adapter Sequence', 'Count Percentage']
+        adapters_df = adapters_df.sort_values(
+            by='Count Percentage',
+            ascending=False,
+        ).reset_index(drop=True)
+        return adapters_df
+
+    def _make_aho_auto(self) -> None:
+        """Add adapter sequences into trie data structure."""
+        LOGGER.debug("Creating trie of adapters")
+        for (adapter, tag) in self.adapters:
+            self.trie.add_word(adapter, (tag, adapter))
+        self.trie.make_automaton()
+
+    def _load_adapters(self) -> None:
```

Comment:
This should include some validation of the sequences and file, e.g., is the file not a text file, are the lines not too long (say, longer than 100 nt), do they lines contain any characters apart from `[acgtnACGTN]`?